### ISSUE TITLE
Can't Query To SlotPos(x,3)

### ISSUE THREAD

**Issue by SjsolisZXA** [SpongePowered/Sponge#1524]
Hello, I can’t seem to be be able to query anything to the the 3rd row of a player’s inventory.

**Test Code:**

```
    player.getInventory().query(new SlotPos(0,3)).set(ItemStack.of(ItemTypes.COOKED_FISH, 1));
    player.getInventory().query(new SlotPos(1,3)).set(ItemStack.of(ItemTypes.DIAMOND, 1));
    
    player.getInventory().query(new SlotPos(0,2)).set(ItemStack.of(ItemTypes.WEB, 1));
    player.getInventory().query(new SlotPos(1,2)).set(ItemStack.of(ItemTypes.APPLE, 1));
    
    player.getInventory().query(new SlotPos(0,1)).set(ItemStack.of(ItemTypes.ACACIA_DOOR, 1));
    player.getInventory().query(new SlotPos(1,1)).set(ItemStack.of(ItemTypes.BAKED_POTATO, 1));
    
    player.getInventory().query(new SlotPos(0,0)).set(ItemStack.of(ItemTypes.ANVIL, 1));
    player.getInventory().query(new SlotPos(1,0)).set(ItemStack.of(ItemTypes.COOKED_RABBIT, 3));
```

**Result:**

![issue](https://user-images.githubusercontent.com/24198315/30334091-3a24816e-97a4-11e7-9481-cef4e7bbcb8d.png)

**SpongeForge:** spongeforge-1.10.2-2281-5.2.0-BETA-2610
**OS:** Windows 10

**Comment by Cybermaxke** [comment]
Can you try `SlotPos.of(0,3)` as `SlotPos`? This uses the `EQUAL` operation instead of the `DELEGATE` one.

**Comment by ustc-zzzz** [comment]
No, it's not the point, although after #1642 you should use `SlotPos#of`.

`SlotPos` is a property added in `GridInventory` (also `Inventory2D`), and there are two `GridInventori`es in player's inventory: the hotbar (including 9 slots), and the main inventory (including 27 slots).

When `Inventory#query` is executed, Sponge will make a depth-first search, and it will not only search the player's inventory, but also search all the child inventories. Therefore, the query result contains 2 slots if the `x` of `SlotPos` is 0. For example, if you queried `SlotPos.of(1, 0)`, the query result contains 2 slots, the first of which is in the hotbar, and the second of which is in the main inventory. Then you called `Inventory#set`, which will only set the first slot (in the hotbar) to a cooked rabbit. The second slot (in the main inventory) is still empty, which should have contained a cooked rabbit.

One solution is using `PlayerInventory#getMain` instead, and another solution is using `SlotIndex` rather than using `SlotPos`. In fact, I think that the current implementation of inventory properties is not completely correct. In my opinion, the query result of an inventory should only contain the inventories whose properties belong to the queried inventory **itself**, and properties belong to those children should be **ignored** in the query result. I'm not quite sure what opinions other developers are holding. @JBYoshi @Faithcaio

**Comment by ustc-zzzz** [comment]
Let me explain further about what's the meaning of "belonging to". Consider a small chest which contains 27 slots. Every slot has a global inventory property called `SlotIndex` *belonging to* the whole chest, from 0 to 26. Besides, a chest inventory has 9 columns and 3 rows. Every slot also has 2 `SlotIndex`es, the one is *belonging to* a column, from 0 to 2, and the other is *belonging to* a row, from 0 to 8.

If somebody wants to query `SlotIndex.of(1)`, every inventory shown above will be searched. Firstly, it will find the 2nd slot, the property of which *belongs to* the chest inventory itself. Secondly, it will find the 2nd, 11th, and 20th slots, the properties of which *belongs to* 3 rows respectively. Thirdly, it will find all the 9 slots from 10th to 18th, the properties of which *belongs to* 9 columns respectively. The query result contains 11 slots distributed like a cross, and in most cases we only want the first slot. Is this the result we really want for operating inventories?

**Comment by ImMorpheus** [comment]
Talked with Faithcaio. This is fixed in 1.12, this can be closed.
